x-pack/plugins/ml/public/url_generator.ts

```diff
@@ -37,22 +109,94 @@ export interface ExplorerUrlState {
    * Optionally set the time range in the time picker.
    */
   timeRange?: TimeRange;
+  /**
+   * Optionally set the refresh interval.
+   */
+  refreshInterval?: RefreshInterval;
+  /**
+   * Optionally set the query.
+   */
+  query?: any;
   /**
    * Optional state for the swim lane
    */
   mlExplorerSwimlane?: ExplorerAppState['mlExplorerSwimlane'];
   mlExplorerFilter?: ExplorerAppState['mlExplorerFilter'];
 }
 
+export interface TimeSeriesExplorerGlobalState {
+  ml: {
+    jobIds: JobId[];
+  };
+  time?: TimeRange;
+  refreshInterval?: RefreshInterval;
+}
+
+export interface TimeSeriesExplorerAppState {
+  zoom?: {
+    from?: string;
+    to?: string;
+  };
+  mlTimeSeriesExplorer?: {
+    detectorIndex?: number;
+    entities?: Record<string, string>;
+  };
+  query?: any;
+}
+
+export interface TimeSeriesExplorerUrlState
+  extends Pick<TimeSeriesExplorerAppState, 'zoom' | 'query'>,
+    Pick<TimeSeriesExplorerGlobalState, 'refreshInterval'> {
+  /**
+   * ML App Page
+   */
+  page: typeof ML_PAGES.TIME_SERIES_EXPLORER;
+  jobIds: JobId[];
+  timeRange?: TimeRange;
+  detectorIndex?: number;
+  entities?: Record<string, string>;
+}
+
+export interface DataFrameAnalyticsQueryState {
+  jobId?: JobId | JobId[];
+  groupIds?: string[];
+}
+
+export interface DataFrameAnalyticsUrlState extends DataFrameAnalyticsQueryState {
+  page: typeof ML_PAGES.DATA_FRAME_ANALYTICS;
+}
+
+export interface DataVisualizerUrlState {
+  page:
+    | typeof ML_PAGES.DATA_VISUALIZER
+    | typeof ML_PAGES.DATA_VISUALIZER_FILE
+    | typeof ML_PAGES.DATA_VISUALIZER_INDEX_SELECT;
+}
+
+export interface DataFrameAnalyticsExplorationQueryState {
+  ml: {
+    jobId: JobId;
+    analysisType: ANALYSIS_CONFIG_TYPE;
+  };
+}
+
+export interface DataFrameAnalyticsExplorationUrlState {
```

Comment:
Nit: Might be worth moving all the types into their own file.

Reply:
Moved to their own files here https://github.com/elastic/kibana/pull/75696/commits/c552644af9106cf2d2f18b434f395dc7631e41e4